mne/epochs.py

```diff
@@ -1342,6 +1342,8 @@ def _get_data(self, out=True, verbose=None):
                         epoch = self._data[idx]
                 else:  # from disk
                     epoch_noproj = self._get_epoch_from_raw(idx)
+                    if epoch_noproj is None:  # Dropped due to bad segment
+                        continue
```

Comment:
To do it, I think you need to refactor quite a bit of the code to make it neat. The other option is to carry the data around, do the detrending, apply the projectors and then reject it, which seems less clean than the solution I have here. The third option is to make the ``drop_log`` always have two values when rejecting by annotation (reason + ``'NO_DATA'``) which also seems less clean.